linera-witty/src/type_traits/implementations/frunk.rs

```diff
@@ -155,3 +165,97 @@ where
+    /// boundary.
+    const SIZE_STARTING_AT_8_BYTE_BOUNDARY: u32;
+
+    /// The size of the list considering the current size calculation starts at the 7-byte
+    /// boundary.
+    const SIZE_STARTING_AT_7_BYTE_BOUNDARY: u32;
+
+    /// The size of the list considering the current size calculation starts at the 6-byte
+    /// boundary.
+    const SIZE_STARTING_AT_6_BYTE_BOUNDARY: u32;
+
+    /// The size of the list considering the current size calculation starts at the 5-byte
+    /// boundary.
+    const SIZE_STARTING_AT_5_BYTE_BOUNDARY: u32;
+
+    /// The size of the list considering the current size calculation starts at the 4-byte
+    /// boundary.
+    const SIZE_STARTING_AT_4_BYTE_BOUNDARY: u32;
+
+    /// The size of the list considering the current size calculation starts at the 3-byte
+    /// boundary.
+    const SIZE_STARTING_AT_3_BYTE_BOUNDARY: u32;
+
+    /// The size of the list considering the current size calculation starts at the 2-byte
+    /// boundary.
+    const SIZE_STARTING_AT_2_BYTE_BOUNDARY: u32;
+
+    /// The size of the list considering the current size calculation starts at the 1-byte
+    /// boundary.
+    const SIZE_STARTING_AT_1_BYTE_BOUNDARY: u32;
+
+    /// The type of the first element of the list, used to determine the current necessary
+    /// alignment.
+    type FirstElement: WitType;
+}
+
+impl SizeCalculation for HNil {
+    const SIZE_STARTING_AT_8_BYTE_BOUNDARY: u32 = 0;
+    const SIZE_STARTING_AT_7_BYTE_BOUNDARY: u32 = 0;
+    const SIZE_STARTING_AT_6_BYTE_BOUNDARY: u32 = 0;
+    const SIZE_STARTING_AT_5_BYTE_BOUNDARY: u32 = 0;
+    const SIZE_STARTING_AT_4_BYTE_BOUNDARY: u32 = 0;
+    const SIZE_STARTING_AT_3_BYTE_BOUNDARY: u32 = 0;
+    const SIZE_STARTING_AT_2_BYTE_BOUNDARY: u32 = 0;
+    const SIZE_STARTING_AT_1_BYTE_BOUNDARY: u32 = 0;
+
+    type FirstElement = ();
+}
+
+macro_rules! calculate_size {
+    ($boundary_offset:expr) => {{
+        let alignment = <<Self::FirstElement as WitType>::Layout as Layout>::ALIGNMENT;
+        let padding = (-($boundary_offset as i32) & (alignment as i32 - 1)) as u32;
+        let size_after_head = padding + Head::SIZE;
+
+        let tail_size = match ($boundary_offset + size_after_head) % 8 {
+            0 => Tail::SIZE_STARTING_AT_8_BYTE_BOUNDARY,
+            1 => Tail::SIZE_STARTING_AT_1_BYTE_BOUNDARY,
+            2 => Tail::SIZE_STARTING_AT_2_BYTE_BOUNDARY,
+            3 => Tail::SIZE_STARTING_AT_3_BYTE_BOUNDARY,
+            4 => Tail::SIZE_STARTING_AT_4_BYTE_BOUNDARY,
+            5 => Tail::SIZE_STARTING_AT_5_BYTE_BOUNDARY,
+            6 => Tail::SIZE_STARTING_AT_6_BYTE_BOUNDARY,
+            7 => Tail::SIZE_STARTING_AT_7_BYTE_BOUNDARY,
+            _ => unreachable!(),
+        };
+
+        size_after_head + tail_size
+    }};
+}
+
+impl<Head, Tail> SizeCalculation for HCons<Head, Tail>
+where
+    Head: WitType,
+    Tail: SizeCalculation,
+{
+    const SIZE_STARTING_AT_8_BYTE_BOUNDARY: u32 = calculate_size!(8);
```

Comment:
Can we make bindings explicit? `calculate_size!(Head, Tail, 8);`

Reply:
Not needed after using the const array.